keps/sig-auth/4317-pod-certificates/README.md

```diff
@@ -0,0 +1,1474 @@
+* The signer name from which a certificate is being requested.
+* (In the status subobject): An issued certificate chain, stored as a single
+  PEM-formatted string.  If there is more than one issued certificate, the first
+  element of the chain is assumed to be the leaf certificate, followed by any
+  necessary intermediates in leaf-to-root order.
+* (In the status subobject): A timestamp at which time the requester that made
+  the PodCertificateRequest should begin trying to refresh the certificate.
+
+PodCertificateRequest objects have a simplified lifecycle compared to
+CertificateSigningRequests.  There are no separate approval and signing stages.
+They can only be issued (and thus implicitly approved), denied (represented by a
+"Denied" status), or failed (represented by a "Failed" status).
+
+Similar to CertificateSigningRequests, controllers that want to issue a
+certificate in response to a PodCertificateRequest need to have two separate
+RBAC permissions:
+* `update` on `podcertificaterequests/status`.
+* `sign` on the requested signer.
+Similar to CertificateSigningRequest, we will also support wildcards on the RBAC signer permissions.
+
+Once a signer controller decides that a PodCertificateRequest is addressed to it
+by matching on the signer name, it should take one of the following actions:
+* Issue the certificate by updating the status subresource of the
+  PodCertificateRequest, setting both the `certificateChain` and
+  `beginRefreshAt` fields.
+* Reject the certificate by updating the status subresource of the
+  PodCertificateRequest, adding a "Denied" condition.
+* Indicate failure in issuing the certificate by updating the status subresource
+  of the PodCertificateRequest, adding a "Failed" condition.
+
+PodCertificateRequest validation logic will:
+* Confirm that the public key is one of the supported key types.
+* Confirm that the proof-of-possession is valid.
+* Confirm that the issued chain (if one is set) consists of valid certificates,
+  that do indeed form a chain to each other.
+* Confirm that neither or both of the certificateChain and beginRefreshAt
+  timestamp are set.
+* Ensure that all fields in PodCertificateRequest.Spec are immutable.
+* Ensure that all fields in PodCertificateRequest.Status are immutable once the
+  request has been issued, denied, or failed.
+
+The `maxExpirationSeconds` field will have the following logic:
+* kube-apiserver will be configurable with an optional maximum duration to allow
+  for certificates issued via PodCertificateRequests, specified in the
+  `--pod-certificate-request-max-duration` flag.  The syntax of this flag will
+  support setting different maximum durations for different signers, with some
+  level of wildcarding.  If a signer is not covered by the flag value,
+  kube-apiserver will assume a maximum duration of 24 hours.
+* If the creator of the PodCertificateRequest left `maxExpirationSeconds` at 0,
+  then kube-apiserver will default it to 24 hours.
```

Comment:
follow-up -> decide if we need to support both a max exp seconds per signer and a default per signer.